src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,932 @@
+    struct dicom_file *f = dicom_file_new(filename, &local_error);
+    if (!f) {
+      g_debug("open failed: %s", local_error->message);
+      g_error_free(local_error);
+      continue;
+    }
+
+    g_hash_table_insert(dicom_file_hash, g_steal_pointer(&filename), f);
+  }
+
+  g_debug("found WSI DICOM files:");
+  g_hash_table_foreach(dicom_file_hash, print_file_from_hash, NULL);
+
+  // the filename we were passed should be one of these WSI files ... get the
+  // slide id from that
+  struct dicom_file *start = g_hash_table_lookup(dicom_file_hash, filename); 
+  if (!start) {
+    g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED, 
+        "%s is not a DICOM WSI", filename);
+    return false;
+  }
+
+  if (!read_whole_file(start)) { 
+    return NULL;
+  }
+
+  const char *slide_id;
+  if (!get_tag_str(start->metadata, SeriesInstanceUID, 0, &slide_id)) {
+    return false;
+  }
+
+  // throw away all files which don't have this slide_id
+  g_hash_table_foreach_remove(dicom_file_hash, 
+                              remove_bad_slide_id, 
+                              (char *) slide_id);
+
+  // pull out the subset of DICOM files that look like pyramid levels
+  g_autoptr(GHashTable) level_hash =
+    g_hash_table_new_full(g_str_hash,
+                          g_str_equal, 
+                          g_free,
+                          (GDestroyNotify) level_destroy);
+  g_hash_table_foreach_steal(dicom_file_hash, find_levels, level_hash);
+
+  g_debug("found pyr levels DICOM files:");
+  g_hash_table_foreach(level_hash, print_level_from_hash, NULL);
+
+  // make the rendering grid
+  g_hash_table_foreach(level_hash, make_grid, osr);
+
+  g_debug("final pyr levels:");
+  g_hash_table_foreach(level_hash, print_level_from_hash, NULL);
+
+  // now sort levels by image_width to make the level array
+  guint level_count = g_hash_table_size(level_hash);
+  if (level_count == 0) {
+    g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                "Couldn't find any tiles");
+    return false;
+  }
+
+  g_autoptr(GPtrArray) level_array = 
+    g_ptr_array_new_full(level_count, (GDestroyNotify) level_destroy);
+  g_hash_table_foreach_steal(level_hash, add_level_array, level_array);
+  g_ptr_array_sort(level_array, compare_level_width); 
+
+  // steal any associated images from the remaining set
+  g_hash_table_foreach_steal(dicom_file_hash, find_associated, osr);
+
+  g_assert(osr->data == NULL);
+  g_assert(osr->levels == NULL);
+
+  // build ops data
+  struct dicom_ops_data *data = g_slice_new0(struct dicom_ops_data);
+  data->dirname = g_strdup(dirname);
+
+  guint count = level_array->len;
+  // we need to use g_ptr_array_free() to steal the array for openslide, but
+  // level_array is an autoptr ... add an extra ref
+  g_ptr_array_ref(level_array);
+  osr->levels = (struct _openslide_level **) g_ptr_array_free(level_array, 
+                                                              false);
```

Comment:
This should just be `g_ptr_array_free(g_steal_pointer(&level_array), false)`.

Reply:
OK, changed.